### DISCUSSION TITLE
How to load the default of a repeater AFTER validation on a wizard form

### DISCUSSION THREAD

**Discussion by StaffCollab** [filamentphp/filament#9915] [Help]
I have a Wizard, and one step includes a repeater field, but the ->default() array on it comes from a livewire field, which changes throughout the form-filling process.
I think the default is being fetched on the form load, even if the repeater is on step three.
Is there a different way to set the repeater children's value so it should be fresh when we arrive at step 3?

I hope I explained it correctly since I'm super tired right now...

**Comment by leandrocfe** [comment]
Could you share the code please?

**Reply by StaffCollab** [reply]
Have you look at my code?

It is a fundamental question: I want to add a $get method into the default so it would calculate on the fly. I cannot just use a $set for it because it is a default for a repeater, and I believe that set wouldn't work on it, or I'm missing something.

**Comment by StaffCollab** [comment]
Here is my code, a bit sloppy as it is still in the works:)

~~~
class PaymentRelationManager extends RelationManager
{

    public ?int $totalAmount = 0;
    public ?array $particulars = [];


    public function form(Form $form): Form
    {
        return AccountPaymentResource::form($form);
    }

    public function table(Table $table): Table
    {
        return AccountPaymentResource::table($table)
            ->headerActions([
                CreateAction::make()->steps([
                    Step::make('Select Charges')
                        ->schema([
                            Section::make([
                                ->schema([
                                    TextInput::make('charge now')
                                        ->reactive(),
                                ])
                                ->addable(false)
                                ->reorderable(false)
                                ->itemLabel(fn (array $state): ?string => 'Advertisement ID - ' . $state['advertisement'] . ' - ' . money($state['charge now'])  ?? null)
                                ->columns(2)
                                ->grid(2)
                                ->collapsible()
                                ->collapsed()
                        ]),
                    Step::make('Payment Plan')
                        ->schema([
                            TextInput::make('installment')->label('installments')->numeric()->default(1),
                            Hidden::make('user')->default(auth()->id()),
                            DatePicker::make('start_date')->label('Start at')->default(today()),
                            Select::make('interval_type')
                                ->relationship('intervalType', 'interval_name'),
                            Select::make('payment_type')
                                ->relationship('paymentType', 'item_name', fn (Builder $query) => $query->where('is_payment', 1)),
                        ])->columns(2)
                        ->afterValidation(function(Get $get){
                            $this->calculateParticulars($get);
                        }),
                    Step::make('Particular Payments')
                        ->schema([
                            Repeater::make('Particulars')
                                ->default($this->particulars)
                                ->schema([
                                    DatePicker::make('payment_date'),
                                    TextInput::make('payment_amount')
                                ]),
                        ]),
                ]),
            ]);
    }
    public function calculateParticulars($get)
    {
        for ($loop = 0; $loop < $get('installment'); $loop++) {

            switch ($get('interval_type')) {
                case 0:
                    $date = Carbon::parse($get('start_date'))->addMonths($loop);
                    break;
                case 1:
                    $date = Carbon::parse($get('start_date'))->addMonths($loop);
                    break;
                case 2:
                    $date = Carbon::parse($get('start_date'))->addWeeks($loop);
                    break;
                case 3:
                    $date = Carbon::parse($get('start_date'))->addWeeks($loop * 2);
                    break;
                case 4:
                    $date = Carbon::parse($get('start_date'))->addDays($loop);
                    break;
            }

            $paymentAmount = $this->totalAmount / $get('installment');

            $this->particulars[] = [
                'payment_date' => $date,
                'payment_amount' => round($paymentAmount, 2),
            ];
        }
    }

   
        }
    }
}
~~~

I have tried to cut unnecessary code, and I hope it is clear enough. 
So basically, in step 3 (particular payments), I want to set a default from a livewire property `$this->particular`, but I need it in the current state because within the `->afterValidation()` of step two, I'm changing the property up.

**Comment by FrazeColder** [comment]
Have you got it working? @StaffCollab I need this one to work as well!

**Comment by lukas-frey** [comment]
Anyone managed this? I have a same issue, I need to call `fill()` (or basically any other way to trigger `default()`) on the whole `Step` component as soon as it is loaded.

**Comment by amimhayden22** [comment]
Has anyone succeeded?

**Comment by lukas-frey** [comment]
Whatever I tried it always failed on an error that the `container` is not initialized yet.

So what I ended up doing was registering a listener for the `wizard::nextStep` form event in the step I wanted and making sure the "next step" is the correct step.

Then I looped through the child items of the Step and loaded the default state using `getDefaultState()` and then overwritten the livewire state manually. So something like `data_set($livewire, '<statepath-of-the-component>', $defaultState)`.

I hoped I could call `fill()` on the Step or something to do this for me, but it always results in an exception that the parent container is not initialized.
